contracts/src/v0.1/L1EndpointRequestResponse.sol

```diff
@@ -0,0 +1,301 @@
+        uint64 accId,
+        uint32 callbackGasLimit,
+        uint8 numSubmission,
+        address sender,
+        uint256 l2RequestId
+    ) public returns (uint256 requestId) {
+        bytes32 jobId = keccak256(abi.encodePacked("bool"));
+        Orakl.Request memory req = buildRequest(jobId);
+        req.add(
+            "get",
+            "https://min-api.cryptocompare.com/data/pricemultifull?fsyms=KLAY&tsyms=USD"
+        );
+        req.add("path", "RAW,KLAY,USD,PRICE");
+        req.add("pow10", "8");
+
+        requestId = requestData(accId, callbackGasLimit, numSubmission, sender, l2RequestId, req);
+    }
+
+    // request for string
+    function requestDataString(
+        uint64 accId,
+        uint32 callbackGasLimit,
+        uint8 numSubmission,
+        address sender,
+        uint256 l2RequestId
+    ) public returns (uint256 requestId) {
+        bytes32 jobId = keccak256(abi.encodePacked("string"));
+        Orakl.Request memory req = buildRequest(jobId);
+        req.add(
+            "get",
+            "https://min-api.cryptocompare.com/data/pricemultifull?fsyms=KLAY&tsyms=USD"
+        );
+        req.add("path", "RAW,KLAY,USD,PRICE");
+        req.add("pow10", "8");
+
+        requestId = requestData(accId, callbackGasLimit, numSubmission, sender, l2RequestId, req);
+    }
+
+    // request for bytes32
+    function requestDataBytes32(
+        uint64 accId,
+        uint32 callbackGasLimit,
+        uint8 numSubmission,
+        address sender,
+        uint256 l2RequestId
+    ) public returns (uint256 requestId) {
+        bytes32 jobId = keccak256(abi.encodePacked("bytes32"));
+        Orakl.Request memory req = buildRequest(jobId);
+        req.add(
+            "get",
+            "https://min-api.cryptocompare.com/data/pricemultifull?fsyms=KLAY&tsyms=USD"
+        );
+        req.add("path", "RAW,KLAY,USD,PRICE");
+        req.add("pow10", "8");
+
+        requestId = requestData(accId, callbackGasLimit, numSubmission, sender, l2RequestId, req);
+    }
+
+    // request for bytes
+    function requestDataBytes(
+        uint64 accId,
+        uint32 callbackGasLimit,
+        uint8 numSubmission,
+        address sender,
+        uint256 l2RequestId
+    ) public returns (uint256 requestId) {
+        bytes32 jobId = keccak256(abi.encodePacked("bytes"));
+        Orakl.Request memory req = buildRequest(jobId);
+        req.add(
+            "get",
+            "https://min-api.cryptocompare.com/data/pricemultifull?fsyms=KLAY&tsyms=USD"
+        );
+        req.add("path", "RAW,KLAY,USD,PRICE");
+        req.add("pow10", "8");
+
+        requestId = requestData(accId, callbackGasLimit, numSubmission, sender, l2RequestId, req);
+    }
+
+    function fulfillDataRequest(uint256 requestId, uint128 response) internal override {
+        bytes32 jobId = keccak256(abi.encodePacked("uint128"));
+        emit DataRequestFulfilled(
+            requestId,
+            sRequest[requestId].l2RequestId,
```

Comment:
Multiple access to `storage` variable is expensive. It is better to copy the data from `storage` to `memory` and then access the struct values. It is possible that compiler can do the job instead of us, but it is better not to rely on it too much.

```Solidity
Request memory r = sRequest[requestId];
emit DataRequestFulfilled(
    requestId,
    r.l2RequestId,
    r.sender,
    r.callbackGasLimit,
...
```

We could also create a pointer to `storage` variable which again would be cheaper than current solution, however, I think because we access all values from struct, it is better to copy them to `memory`.